dedupe/convenience.py

```diff
@@ -24,55 +24,88 @@ def consoleLabel(deduper): # pragma: no cover
-                  file=sys.stderr)
-            print('Do these records refer to the same thing?', file=sys.stderr)
-            valid_response = False
-            while not valid_response:
+    while not finished :       
+        if len(record_label_buffer) and use_previous:
+            uncertain_pairs.append(record_label_buffer[0][0]) # append previous record_pair
+            record_label_buffer = record_label_buffer[1:] # remove previous pair from buffer
+        use_previous = False
+        
+        if not uncertain_pairs:
+            uncertain_pairs = deduper.uncertainPairs()
+        record_pair = uncertain_pairs[-1]
+                
+        n_match_in_buffer = sum(x[1]=='match' for x in record_label_buffer) 
+        n_distinct_in_buffer = sum(x[1]=='distinct' for x in record_label_buffer)
+                                                                                        
+        n_match, n_distinct = (len(deduper.training_pairs['match']) + n_match_in_buffer,
+                        len(deduper.training_pairs['distinct']) + n_distinct_in_buffer)
+      
+        label = ''
+
+        for pair in record_pair:
+            for field in fields:
+                line = "%s : %s" % (field, pair[field])
+                print(line, file=sys.stderr)
+            print(file=sys.stderr) 
+
+        print("{0}/10 positive, {1}/10 negative".format(n_match, n_distinct),
+                file=sys.stderr)
+        print('Do these records refer to the same thing?', file=sys.stderr)
+        valid_response = False
+        while not valid_response:
+            if len(record_label_buffer):
+                print('(y)es / (n)o / (u)nsure / (f)inished / (p)revious', file=sys.stderr)
+                label = input()
+                if label in ['y', 'n', 'u', 'f', 'p']:
+                    valid_response = True
+            else:
                 print('(y)es / (n)o / (u)nsure / (f)inished', file=sys.stderr)
                 label = input()
                 if label in ['y', 'n', 'u', 'f']:
                     valid_response = True
-
-            if label == 'y' :
-                examples['match'].append(record_pair)
-                labeled = True
-            elif label == 'n' :
-                examples['distinct'].append(record_pair)
-                labeled = True
-            elif label == 'f':
-                print('Finished labeling', file=sys.stderr)
-                finished = True
-            elif label != 'u':
-                print('Nonvalid response', file=sys.stderr)
-                raise
-
-        if labeled :
-            deduper.markPairs(examples)
+                if label == 'p':
+                    print('No record in memory: cannot use (p)revious', file=sys.stderr)
+
+        if label == 'y' :
+            record_label_buffer.insert(0, (record_pair, 'match'))
+            uncertain_pairs.pop() # Remove current pair form uncertain list
```

Comment:
If you pop above, when you get record_pair you don't need all these pops down here.